plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs

```diff
@@ -187,3 +198,44 @@ instance (GShow uni, GEq uni, uni `Includes` Integer) => ToBuiltinMeaning uni Ex
                 :: a ~ Opaque term (TyVarRep ('TyNameRep "a" 0))
                 => Void -> a)
             (\_ _ -> ExBudget 1 0)
+    toBuiltinMeaning Null = makeBuiltinMeaning nullPlc mempty where
+        nullPlc :: SomeValueN uni [] '[a] -> Bool
+        nullPlc (SomeValueArg _ (SomeValueRes _ xs)) = null xs
+    toBuiltinMeaning Head = makeBuiltinMeaning headPlc mempty where
+        headPlc :: SomeValueN uni [] '[a] -> EvaluationResult (Opaque term a)
+        headPlc (SomeValueArg uniA (SomeValueRes _ xs)) = case xs of
+            x : _ -> EvaluationSuccess . Opaque . fromConstant $ someValueOf uniA x
+            _     -> EvaluationFailure
+    toBuiltinMeaning Tail = makeBuiltinMeaning tailPlc mempty where
+        tailPlc :: SomeValueN uni [] '[a] -> EvaluationResult (SomeValueN uni [] '[a])
+        tailPlc (SomeValueArg uniA (SomeValueRes uniListA xs)) = case xs of
+            _ : xs' -> EvaluationSuccess . SomeValueArg uniA $ SomeValueRes uniListA xs'
+            _       -> EvaluationFailure
+    toBuiltinMeaning Fst = makeBuiltinMeaning fstPlc mempty where
+        fstPlc :: SomeValueN uni (,) '[a, b] -> Opaque term a
+        fstPlc (SomeValueArg uniA (SomeValueArg _ (SomeValueRes _ (x, _)))) =
+            Opaque . fromConstant . Some $ ValueOf uniA x
+    toBuiltinMeaning Snd = makeBuiltinMeaning sndPlc mempty where
+        sndPlc :: SomeValueN uni (,) '[a, b] -> Opaque term b
+        sndPlc (SomeValueArg _ (SomeValueArg uniB (SomeValueRes _ (_, y)))) =
+            Opaque . fromConstant . Some $ ValueOf uniB y
+    toBuiltinMeaning Swap = makeBuiltinMeaning swapPlc mempty where
+        swapPlc :: SomeValueN uni (,) '[a, b] -> SomeValueN uni (,) '[b, a]
```

Comment:
Made inference work with polymorphic built-in types just like it works with regular type variables!